### PR TITLE
Super class field mapping Resource Extraction

### LINKED ISSUE #943: Handle inherited properties of Resource into questionnaire when doing extraction via ResourceMapper
**Describe the bug**
Fhirpath mapping into questionnaire allow mapping definition like below
 "definition": "http://hl7.org/fhir/StructureDefinition/Patient#Patient.name.given"

ResourceMapper processes it an automatically maps the questionnaire-response answer to mapping value. But this work only for declared properties. For inherited properties (from DomainResource, and REsource like tag, extension) it does not work 

**Component**
SDC library

**To Reproduce**
SDK has an issue with the way it works for fhir-path
```
private fun Class<*>.getFieldOrNull(name: String): Field? {
  return try {
    getDeclaredField(name)
  } catch (ex: NoSuchFieldException) {
    return null
  }
}
```
Instead of picking declared-field we should also pick fields inherited. Patient or any other resource extends DomainResource and the Resource. It means that we can not set inherited fields like meta.tag/extension etc using fhir-path.

### PR DESCRIPTION
**Pull request by maimoonak**
Fixes #943 

**Description**
Resource Extraction by Field allows mapping for parent class fields like DomainResource field extension for Patient class

**Alternative(s) considered**
The class just needed additional handling to traverse super class when Field not found in current class mapped

**Type**
Choose one: Bug fix

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

**Comment by Tarun-Bhardwaj** [comment]
@aditya-07 , could you please review this PR?

**Comment by maimoonak** [comment]
@Tarun-Bhardwaj @aditya-07 I have fixed the failing CI issue

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 1]
```diff
@@ -607,7 +619,7 @@ private fun Class<*>.getFieldOrNull(name: String): Field? {
   return try {
     getDeclaredField(name)
   } catch (ex: NoSuchFieldException) {
-    return null
+    return superclass.getFieldOrNull(name)
```
Comment (aditya-07) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:622]: superclass is nullable and may cause `NullPointerException` incase the field is not found throughout the class hierarchy.
```suggestion
    return superclass?.getFieldOrNull(name)
```

**Review by aditya-07 — CHANGES_REQUESTED** [review]

**Comment by maimoonak** [comment]
@aditya-07 feedback resolved.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 2]
```diff
@@ -607,7 +619,7 @@ private fun Class<*>.getFieldOrNull(name: String): Field? {
   return try {
     getDeclaredField(name)
   } catch (ex: NoSuchFieldException) {
-    return null
+    return superclass?.getFieldOrNull(name)
```
Comment (epicadk) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:622]:
```suggestion
    superclass?.getFieldOrNull(name)
```

**Review by aditya-07 — APPROVED** [review]
Great PR @maimoonak! LGTM!

**Comment by jingtang10** [comment]
@maimoonak @aditya-07: Looking through this code, I understand that the change you made is making sure a latter value is not overriding an earlier value if you're setting the same field (which is a list).

So I'm not sure the title of this PR is accurate - can you confirm? and if so can you update the title and description?

thanks

**Comment by maimoonak** [comment]
@jingtang10 The main aim of PR as to allow mapping fields into superclass. For example with 
`Patient->DomainResource->Resource`
code did not allow to map tag OR extension fields into superlcass to be mapped into SDC-extraction. This PR fixes it by [line](https://github.com/google/android-fhir/pull/954/files#diff-4d2985aba5e8203b97cc5dc2eac7dbbd0c7832be34a27036d1ecfb8e4a3a3e50R622). 
Other issue which was noticed when mapping extension was that incase of multiple mapping it was always overriding first value of extension hence only last mapped extension persisted. 

The issue that was fixed into PR was 'Allow super class field mapping for SDC Resource Extraction'

**Comment by jingtang10** [comment]
thanks @maimoonak - as long as this is documented in this issue it's fine. if you can just list the two things this PR does in the description that'd be great. thanks!

**Comment by jingtang10** [comment]
But perhaps a good follow-up PR to write is to provide minimal test cases. I think your new test case is too big and testing both of the things this PR is doing.

**Comment by maimoonak** [comment]
Ok. Sure would break it down into two separate cases for each functionality separately

**Pull request merged**